portable/GCC/ATmegaxxxx/port.c

```diff
@@ -0,0 +1,822 @@
+                                "pop    r12                                     \n\t"   \
+                                "pop    r11                                     \n\t"   \
+                                "pop    r10                                     \n\t"   \
+                                "pop    r9                                      \n\t"   \
+                                "pop    r8                                      \n\t"   \
+                                "pop    r7                                      \n\t"   \
+                                "pop    r6                                      \n\t"   \
+                                "pop    r5                                      \n\t"   \
+                                "pop    r4                                      \n\t"   \
+                                "pop    r3                                      \n\t"   \
+                                "pop    r2                                      \n\t"   \
+                                "pop    __zero_reg__                            \n\t"   \
+                                "pop    __tmp_reg__                             \n\t"   \
+                                "out    __SREG__, __tmp_reg__                   \n\t"   \
+                                "pop    __tmp_reg__                             \n\t"   \
+                             );
+#endif
+/*-----------------------------------------------------------*/
+
+/*
+ * Perform hardware setup to enable ticks from relevant Timer.
+ */
+static void prvSetupTimerInterrupt( void );
+/*-----------------------------------------------------------*/
+
+/*
+ * See header file for description.
+ */
+StackType_t *pxPortInitialiseStack( StackType_t *pxTopOfStack, TaskFunction_t pxCode, void *pvParameters )
+{
+uint16_t usAddress;
+
+    /* Place a few bytes of known values on the bottom of the stack.
+    This is just useful for debugging. */
+
+    *pxTopOfStack = 0x11;
+    pxTopOfStack--;
+    *pxTopOfStack = 0x22;
+    pxTopOfStack--;
+    *pxTopOfStack = 0x33;
+    pxTopOfStack--;
+
+    /* Simulate how the stack would look after a call to vPortYield() generated by
+    the compiler. */
+
+    /* The start of the task code will be popped off the stack last, so place
+    it on first. */
+    usAddress = ( uint16_t ) pxCode;
+    *pxTopOfStack = ( StackType_t ) ( usAddress & ( uint16_t ) 0x00ff );
+    pxTopOfStack--;
+
+    usAddress >>= 8;
+    *pxTopOfStack = ( StackType_t ) ( usAddress & ( uint16_t ) 0x00ff );
+    pxTopOfStack--;
+
+#if defined(__AVR_3_BYTE_PC__)
+    /* The AVR ATmega2560/ATmega2561 have 256KBytes of program memory and a 17-bit
+     * program counter. When a code address is stored on the stack, it takes 3 bytes
+     * instead of 2 for the other ATmega* chips.
+     *
+     * Store 0 as the top byte since we force all task routines to the bottom 128K
+     * of flash. We do this by using the .lowtext label in the linker script.
```

Comment:
The linker script is provided by [binutils-avr](https://ubuntu.pkgs.org/20.04/ubuntu-universe-amd64/binutils-avr_2.26.20160125+Atmel3.6.2-1_amd64.deb.html) used with gcc. This PR is based on solely using the Open Source tools available for AVR. Nothing to do with Atmel Studio (although it would be doing the same thing under the covers, since it uses avr-libc too).

Look for `/usr/lib/avr/lib/ldscripts/avr6.x` in the [`binutils-avr`](https://ubuntu.pkgs.org/20.04/ubuntu-universe-amd64/binutils-avr_2.26.20160125+Atmel3.6.2-1_amd64.deb.html) package.